mage_integrations/mage_integrations/sources/datadog/streams/__init__.py

```diff
@@ -0,0 +1,290 @@
+
+    def get_payload(self, cursor=None, **kwargs) -> Mapping[str, Any]:
+        query = self.config.get('query', {}).get(self.TABLE)
+        if query:
+            kwargs['query'] = query
+        payload = {
+            "filter": kwargs,
+            "page": {"limit": self.max_records_per_request},
+        }
+        if cursor:
+            payload["page"]["cursor"] = cursor
+
+        return payload
+
+
+class AuditLogs(IncrementalSearchableStream):
+    """
+    https://docs.datadoghq.com/api/latest/audit/#search-audit-logs-events
+    """
+    TABLE = 'audit_logs'
+    ENTITY = 'audit_log'
+    API_METHOD = 'POST'
+    SCHEMA = 'audit_logs'
+
+    def get_url(self):
+        return 'https://api.datadoghq.com/api/v2/audit/events/search'
+
+
+class Logs(IncrementalSearchableStream):
+    """
+    https://docs.datadoghq.com/api/latest/logs/#search-logs
+    """
+    TABLE = 'logs'
+    ENTITY = 'log'
+    API_METHOD = 'POST'
+    SCHEMA = 'logs'
+
+    def get_url(self):
+        return 'https://api.datadoghq.com/api/v2/logs/events/search'
+
+
+class BasedListStream(DatadogStream):
+    API_METHOD = 'GET'
+    parse_response_root: Optional[str] = "data"
+
+
+class Metrics(BasedListStream):
+    """
+    https://docs.datadoghq.com/api/latest/metrics/#get-a-list-of-metrics
+    """
+    TABLE = 'metrics'
+    ENTITY = 'metric'
+    API_METHOD = 'GET'
+    SCHEMA = 'metrics'
+
+    def get_url(self):
+        return 'https://api.datadoghq.com/api/v2/metrics'  # max value allowed (2 weeks)
+
+
+class PaginatedBasedListStream(BasedListStream):
+    primary_key: Optional[str] = "id"
+
+    def next_page_token(self, response: Dict) -> Optional[Mapping[str, Any]]:
+        next_offset = response.get("meta", {}).get("pagination", {}).get("next_offset", -1)
+        current_offset = response.get("meta", {}).get("pagination", {}).get("offset", -1)
+        next_page_token = {}
+        if next_offset != current_offset:
+            next_page_token = {"page[offset]": next_offset}
+        return next_page_token
+
+
+class Incidents(PaginatedBasedListStream):
+    """
+    https://docs.datadoghq.com/api/latest/incidents/#get-a-list-of-incidents
+    """
+    TABLE = 'incidents'
+    ENTITY = 'incident'
+    API_METHOD = 'GET'
+    SCHEMA = 'incidents'
+
+    def get_url(self):
+        return 'https://api.datadoghq.com/api/v2/incidents'
+
+
+class IncidentTeams(PaginatedBasedListStream):
+    """
+    https://docs.datadoghq.com/api/latest/incident-teams/#get-a-list-of-all-incident-teams
+    """
+    TABLE = 'incident_teams'
+    ENTITY = 'incident_team'
+    API_METHOD = 'GET'
+    SCHEMA = 'incident_teams'
+
+    def get_url(self):
+        return 'https://api.datadoghq.com/api/v2/teams'
+
+
+class Users(PaginatedBasedListStream):
+    """
+    https://docs.datadoghq.com/api/latest/users/#list-all-users
+    """
+    TABLE = 'users'
+    ENTITY = 'user'
+    API_METHOD = 'GET'
+    SCHEMA = 'users'
+
+    current_page = 0
+
+    def get_url(self):
```

Comment:
nit: if base url is same for all streams, you can specify a base_url constant and only customize path for each stream.